Promote IAC docs for agents and dynamic resources

ptgott opened the pull request:
Closes #27382
Closes #25418
Closes #25442

Two aspects of setting up a Teleport cluster are omnipresent in the docs but have not received dedicated treatment:

- Running agents
- Applying dynamic resources

As a result, it is difficult to include discussions about those topics that are separate from a specific workflow or how-to guide. One glaring absence has been prominent guidance on using infrastructure-as-code tools to achieve these tasks.

This change improves the visibility of Teleport's support for infrastructure-as-code tools by:

- Creating docs sections for running agents and applying dynamic resources
- Making IAC instructions prominent within these sections

The intention is for readers to become familiar with different methods of applying dynamic resources and running agents, including how to do this with IAC, so they can apply this knowledge when reading other parts of the docs.

The new section on dynamic resources also gives us a place to put other guides to using the Terraform provider and Kubernetes Operator, e.g., if we add guides to using these tools with popular GitOps platforms.

Likewise, the section on agents gives us a place to put other agent-wide information, e.g., how to enable an additional Teleport service on an instance that is already running.

## Substantial changes in response to feedback:

### Move the "Dynamic Configuration" section

Make this a subsection of "Manage your Cluster" since it's not
self-evidently clear as a top-level docs section. Users will probably need an introduction via the "Manage your Cluster" section intro page.

This also reverts some of the more drastic sidebar changes introduced by the previous commit.

Responds to xinding33 feedback.

###   Make IAC learning tracks prominent/hard to avoid

Closes #27751

Responds to xinding33 feedback

The motivation is to be more opinionated about the course that users
take through the docs. We currently recommend two tracks for
self-service users, i.e., the users expected to make use of the landing
page:

- Setting up a toy self-hosted Teleport cluster
- Setting up a Teleport Team/Enterprise Cloud cluster that can
      eventually become production ready

By moving the "Get Started" guide to the landing page, we direct users
immediately on to the first track. This change then gives new users a
way to enter the second track from the docs landing page with a
prominent link to the Teleport Team docs.

This change also edits The landing page to direct users who have
completed the getting started experience to instructions for setting up
a pool of agents on Terraform, helping to make infrastructure-as-code a
first-class citizen of the docs.

This change also removes the menu of links that used to confront users
on the landing page. Since all sidebar sections include introduction
pages, users interested in the content of a sidebar section can visit
the section. By removing links, we make it clearer for users how to
proceed through the docs.

ptgott committed:
Promote IAC docs for agents and dynamic resources

Closes #27382
Closes #25418
Closes #25442

Two aspects of setting up a Teleport cluster are omnipresent in the docs
but never get dedicated treatment:

- Running agents
- Applying dynamic resources

As a result, it is difficult to include discussions about those topics
that are separate from a specific workflow or how-to guide. One glaring
absence has been prominent guidance on using infrastructure-as-code
tools to achieve these tasks.

This change improves the visibility of Teleport's support for
infrastructure-as-code tools by:

- Creating top-level docs sections for running agents and applying
  dynamic resources
- Making IAC instructions prominent within these sections

The intention is for readers to become familiar with different methods
of applying dynamic resources and running agents, including how to do
this with IAC, so they can apply this knowledge when reading other parts
of the docs.

As a result, in the docs sidebar, the "Dynamic Resources" section comes
before the section on RBAC, and the "Teleport Agents" section comes
before the sections related to individual Teleport services
("Application Access", "Server Access", etc.).

The new section on dynamic resources also gives us a place to put other
guides to using the Terraform provider and Kubernetes Operator, e.g., if
we add guides to using these tools with popular GitOps platforms.

Likewise, the section on agents gives us a place to put other agent-wide
information, e.g., how to enable an additional Teleport service on an
instance that is already running.

To allow for these changes, I have also made the following, more
tangentially related sidebar changes:

- **Renamed sidebar sections to be noun phrases instead of verb
  phrases:** Currently, one half of the sidebar is made of imperative
  phrases like "Manage Access" and "Manage your Cluster". This doesn't
  really work for the sections on agents and dynamic resources, so I
  have renamed these sections for consistency.

- **Moved the "Manage your Cluster" and "Deploy a Cluster" sections**: I
  have arranged the sidebar so more basic topics (i.e., those that new
  users and experienced users alike will need to be familiar with) are
  on the top and more advanced ones are on the bottom.

  The other sections that are currently in the first half of the sidebar
  are topics that all Teleport users will need to get familiar with,
  while day two operations and self-hosted production deployments are
  more advanced topics.

- **Edited the landing page:** I have edited the landing page of the
  docs to reflect the new sidebar organization. This also makes the page
  shorter, simpler, and more opinionated. It spells out a high-level
  sequence for setting up Teleport, then provides a list of advanced
  topics for further reading.

  Links correspond to sidebar sections--as before, I wanted to describe
  the topic of each sidebar section so users would know this information
  without having to navigate away from the landing page.

ptgott commented:
TODO:

- [ ] Make sure I have added redirects

ptgott committed:
Move the "Dynamic Configuration" section

Make this a subsection of "Manage your Cluster" since it's not
self-evidently clear as a top-level docs section. Users will probably
need an introduction via the "Manage your Cluster" section intro page.

This also reverts some of the more drastic sidebar changes introduced by
the previous commit.

Responds to xinding33 feedback.

ptgott committed:
Fix linter errors

ptgott committed:
Make IAC learning tracks prominent/hard to avoid

Closes #27751

Responds to xinding33 feedback

The motivation is to be more opinionated about the course that users
take through the docs. We currently recommend two tracks for
self-service users, i.e., the users expected to make use of the landing
page:

- Setting up a toy self-hosted Teleport cluster
- Setting up a Teleport Team/Enterprise Cloud cluster that can
  eventually become production ready

By moving the "Get Started" guide to the landing page, we direct users
immediately on to the first track. This change then gives new users a
way to enter the second track from the docs landing page with a
prominent link to the Teleport Team docs.

This change also edits The landing page to direct users who have
completed the getting started experience to instructions for setting up
a pool of agents on Terraform, helping to make infrastructure-as-code a
first-class citizen of the docs.

This change also removes the menu of links that used to confront users
on the landing page. Since all sidebar sections include introduction
pages, users interested in the content of a sidebar section can visit
the section. By removing links, we make it clearer for users how to
proceed through the docs.

ptgott committed:
Incorporate Trivy recommendations

alexfornuto commented on docs/pages/agents/deploy-agents-terraform.mdx:
```diff
@@ -0,0 +1,310 @@
+  proxy user traffic.
+- **Static configuration settings:** Edit the configuration of each service to
+  point to specific resources in your infrastructure.
+
```
Suggest a new paragraph here pointing to a page for configuring static settings and declaring that the next section of the guide will cover dynamic resources.

alexfornuto replied:
Or, is the "Configure Teleport services in the agent pool" section is the counter-point to the "Define dynamic resources in Terraform" section, suggest rephrasing the heading to better clarify that each of the sections relates to each of these bullet points.

ptgott replied:
I've removed the bulleted list to go straight into the H3s—hopefully that clears things up

alexfornuto commented on docs/pages/agents/deploy-agents-terraform.mdx:
```diff
@@ -0,0 +1,310 @@
+|Application|`teleport_app`|
+|Database|`teleport_database`|
+
+To declare a dynamic resource with Terraform, add configuration blocks similar
+to the ones below. The Teleport Terraform provider creates these on the Auth
```
Suggest reminding the reader what file these declarations should be added to.

alexfornuto commented on docs/pages/agents/deploy-agents-terraform.mdx:
````diff
@@ -0,0 +1,310 @@
+<TabItem label="Database">
+
+```text
+resource "teleport_database" "example" {
+    metadata = {
+        name = "example"
+        description = "Test database"
+        labels = {
+            // Teleport adds this label by default, so add it here to
+            // ensure a consistent state.
+            "teleport.dev/origin" = "dynamic"
+        }
+    }
+
+    spec = {
+        protocol = "postgres"
+        uri = "localhost"
+    }
+}
````
Suggest adjusting the tabs in this tab to match the first.

alexfornuto commented on docs/pages/management/dynamic-resources.mdx:
````diff
@@ -0,0 +1,277 @@
+which is equivalent to the two roles shown above:
+
+```yaml
+
````
```suggestion
```

alexfornuto commented on docs/pages/management/introduction.mdx:
```diff
@@ -7,14 +7,20 @@ layout: tocless-doc
+cluster as your infrastructure changes. Teleport provides three methods for
+applying dynamic resources: the `tctl` client tool, Teleport Terraform provider,
+and Kubernetes Operator. Read [Using Dynamic
+Resources](./dynamic-resources.mdx).
```
```suggestion
Resources](./dynamic-resources.mdx) for more information.
```

or 

```suggestion
Resources](./dynamic-resources.mdx) for details.
```

alexfornuto commented on docs/pages/index.mdx:
```diff
@@ -1,175 +1,220 @@
 </Notice>
 
-## Try out Teleport
+- A Linux host with only port `443` open to ingress traffic. You must be able
```
Does LE no longer require port `80` to complete its validation prior to issuing a cert?

ptgott replied:
Teleport uses the TLS-ALPN-01 challenge, which doesn't require port 80

alexfornuto commented on docs/pages/index.mdx:
```diff
@@ -1,175 +1,220 @@
+  </Admonition>
 
-## Choose an edition
+- A two-factor authenticator app such as [Authy](https://authy.com/download/),
+  [Google Authenticator](https://www.google.com/landing/2step/), or [Microsoft
+  Authenticator](https://www.microsoft.com/en-us/account/authenticator)
```
Consider adding 1Password to the list.

alexfornuto commented on docs/pages/index.mdx:
```diff
@@ -1,175 +1,220 @@
-infrastructure. Teleport has four editions: 
+You must also have one of the following:
+- A registered domain name.
+- An authoritative DNS nameserver managed by your organization, plus an existing
+  certificate authority. If using this approach, ensure that your browser is
+  configured to use your organization's nameserver.
```
"plus an existing certificate authority" - Is this required when using LE?

ptgott replied:
LE would be the first option. I've bolded "one" to emphasize that it's an exclusive OR.

alexfornuto commented on docs/pages/index.mdx:
```diff
@@ -1,175 +1,220 @@
+Generate a configuration file for Teleport using the `teleport configure` command.
+This command requires information about a TLS certificate and private key.
+
+(!docs/pages/includes/tls-certificate-setup.mdx!)
```
Just a suggestion: For parity, the "Private network deployment" tab in this partial could be updated to include an example of the relevant generated config, as the first tab does.

ptgott replied:
I've suggested removing the sample config from the "Let's Encrypt" tab. That example config is a little incorrect/misleading anyway, since it doesn't show the `cluster_name` field and only includes the Proxy Service config.

ptgott commented on docs/pages/agents/introduction.mdx:
```diff
@@ -0,0 +1,99 @@
+They should be the only clients that can access a resource without Teleport.
+
+In this setup, agents dial the Teleport Proxy Service in order to establish
+reverse SSH tunnels. While the Proxy Service remains open to the public internet
```
I've clarified this a bit in the next paragraph

ptgott commented on docs/pages/agents/introduction.mdx:
```diff
@@ -0,0 +1,99 @@
+    db["Self-hosted database"]
+    subgraph agent2["Teleport agent"]
+      service3["Teleport Database Service"]
+      service4["Teleport Application Service"]
```
I thought there wouldn't be enough space for this, but it turns out that there is—added.

ptgott commented on examples/agent-pool-terraform/userdata.sh:
EC2 cloud-init scripts log to a file by default:

> The cloud-init output log file captures console output so it is easy to debug your scripts following a launch if the instance does not behave the way you intended. To view the log file, [connect to the instance](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/AccessingInstances.html) and open /var/log/cloud-init-output.log.

(https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/user-data.html)

ptgott commented on docs/pages/agents/deploy-agents-terraform.mdx:
```diff
@@ -0,0 +1,310 @@
+configuration file to the default location, `/etc/teleport.yaml`. The
+configuration file enables each Teleport service we associated with our token.
+
+So we you access hosts in the agent pool later, the configuration also adds the
```
Oof, I mangled that sentence and didn't catch it during editing. I've added a fix

ptgott committed:
Respond to alexfornuto feedback

ptgott committed:
Restore list of dynamic resources to the reference

alexfornuto reviewed (approved):
👍 , though my local render fails on `deploy-agents-terraform` with a `Expected a closing tag for `<TabItem>` error that I couldn't track down the source of.

ptgott committed:
Fix linter warnings

ptgott committed:
Remove the ".sh" extension from userdata script

The Terraform module that reads this file does not need the extension,
which was causing trouble for our shellcheck linter.
